experimental/traffic-portal/src/app/api/testing/topology.service.ts

```diff
@@ -0,0 +1,155 @@
+/*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+import { Injectable } from "@angular/core";
+import {
+	RequestTopology,
+	ResponseTopology,
+	ResponseTopologyNode
+} from "trafficops-types";
+
+import { TopologyService as ConcreteService, TopTreeNode } from "src/app/api";
+
+/**
+ * TopologyService expose API functionality relating to Topologies.
+ */
+@Injectable()
+export class TopologyService {
+	private readonly topologies: ResponseTopology[] = [
+		{
+			description: "",
+			lastUpdated: new Date(),
+			name: "test",
+			nodes: [
+				{
+					cachegroup: "Edge",
+					parents: [1],
+				},
+				{
+					cachegroup: "Mid",
+					parents: [2],
+				},
+				{
+					cachegroup: "Origin",
+					parents: [],
+				},
+			],
+		},
+	];
+
+	/**
+	 * Gets one or all Topologies from Traffic Ops
+	 *
+	 * @param name The unique name of a single Topology to be returned
+	 * @returns Either an Array of Topologies or a single Topology, depending on whether `name` was
+	 * passed.
+	 */
+	public async getTopologies(name?: string): Promise<Array<ResponseTopology> | ResponseTopology> {
+		if (name !== undefined) {
+			const topology = this.topologies.find(t => t.name === name);
+			if (!topology) {
+				throw new Error(`no such Topology ${name}`);
+			}
+			return topology;
+		}
+		return this.topologies;
+	}
```

Comment:
The type of this testing service method doesn't match the type of the method by the same name on the concrete service. That method can only return arrays of Topologies, which means that if you drop this service in as a replacement for that concrete service, things calling `getTopologies` with a specified `name` argument will expect to get an Array back, but actually you've given them a single Topology object.

I'd suggest fixing this not by modifying the testing service, but by changing the concrete service to have an overload signature wherein it returns a single Topology if asked for it by name, i.e. `(name: string) => Promise<ResponseTopology>`

Reply:
I have a strong preference to not change the concrete service this late in the review cycle. Matching the call signature in a5a39eca00